### ISSUE TITLE
LARS/LARC with Horovod in Keras

### ISSUE THREAD

**Issue by JohnTaylor2000** [horovod/horovod#983]
**Environment:**
1. Framework: (TensorFlow, Keras, PyTorch, MXNet)
2. Framework version:eras
3. Horovod version:all
4. MPI version:all
5. CUDA version:all
6. NCCL version:all
7. Python version:all
8. OS and version:all

**Checklist:**
1. Did you search issues to find if somebody asked this question before?
2. If your question is about hang, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/running.md)?
3. If your question is about docker, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/docker.md)?

**Your question:**
Please ask your question here.

Scaling to large batch sizes and large numbers of GPUs without loss of performance is challenging. LARS/LARC is potentially a key requirement for maintaining performance at large batch sizes. It would be very helpful if working example of LARS/LARC in Keras integrated with horovod could be posted as one of the examples please. It would probably also be helpful to add a section about the methods to overcome loss of performance when scaling to large  batch sizes on the main horovod page.

**Comment by alsrgv** [comment]
@JohnTaylor2000, this is an excellent idea!  We'll try to get to it this month, but in case you have time contributions are very welcome.

**Comment by JohnTaylor2000** [comment]
FYI - There is a version of LARS for Keras located here:-

https://github.com/keras-team/keras-contrib/blob/master/keras_contrib/optimizers/lars.py

although I am unsure whether this can be successfully combined with Horovod. LARC also needs to be added.

**Comment by khayamgondal** [comment]
@alsrgv @JohnTaylor2000 Is anyone working on this If not I can take it.

**Comment by JohnTaylor2000** [comment]
That would be great - let me know and I will try it out!

Best wishes,

John


On 31 Aug 2019, at 1:30 pm, Khayam Gondal <notifications@github.com<mailto:notifications@github.com>> wrote:


@alsrgv<https://github.com/alsrgv> @JohnTaylor2000<https://github.com/JohnTaylor2000> Is anyone working on this If not I can take it.

**Comment by tgaddair** [comment]
Thanks @khayamgondal, looking forward to the contribution!

**Comment by tgaddair** [comment]
Hey @khayamgondal, you should be able to wrap the existing LARS optimizer that @JohnTaylor2000 provided in a `hvd.DIstributedOptimizer`, which will do the work of overriding `get_gradients` for you.  So in the example script you're writing, you would just need to do:

```
opt = LARS(...)
opt = hvd.DIstributedOptimizer(opt, ...)
```

**Comment by JohnTaylor2000** [comment]
Hey @khayamgondal<https://github.com/khayamgondal>, Yes, I have tested this quickly and believe that it works, though further testing to verify that this is what happened is required. I have not implemented or tried LARC.

**Comment by khayamgondal** [comment]
Thanks John and Travis, I'll work on it.

**Comment by wuchangping** [comment]
Hi @khayamgondal  @JohnTaylor2000  , how about this featue? Is LARS ready at Horovod?

**Comment by xiaocaijizzz** [comment]
Hi @tgaddair @khayamgondal @JohnTaylor2000 Has anyone tried LARS optimizer at Horovod with AdaSum? I use HOROVOD_GPU_OPERATIONS=NCCL flag to compile Horovod. According to the official statement of Hovorod, only averaging through NCCL library is used to perform reductions and no AdaSum algorithm will take place when training is run on a single node. This means I will get the same result if the training is run on a single node, no matter which op I set, AdaSum or Average. 
And it is true when I take SGD optimizer. But when I use LARS optimizer, even if the training is run on a single node and the seed is fixed, I get quite different results when I set the distribution op as AdaSum and Average, the same results when I use Lamb, Adam, Adamw.
So I'd like to ask if anyone has verified AdaSum at Hovorod with Adam/Adamw/LARS/Lamb optimizer.
